File: halgos/src/Searching.hs
module Searching (linear, binary, ternary) where

linearAux :: (Eq a) => [a] -> a -> Int -> Maybe Int
linearAux [] _ _ = Nothing
linearAux (curitem : items) item index
  | curitem == item = Just index
  | otherwise = linearAux items item (index + 1)

-- | The `linear` method performs linear search on a list for an item, returning an index if found, or `Nothing` if not.
linear :: (Eq a) => [a] -> a -> Maybe Int
linear items item = linearAux items item 0

binaryAux :: (Ord a) => [a] -> a -> Int -> Int -> Maybe Int
binaryAux [] _ _ _ = Nothing
binaryAux items item l r
  | l > r = Nothing
  | curitem == item = Just mid
  | item < curitem = binaryAux items item l (mid - 1)
  | item > curitem = binaryAux items item (mid + 1) r
  where
    mid = div (l + r) 2
    curitem = items !! mid

-- | The `binary` method performs binary search on a list for an item, returning an index if found, or `Nothing` if not.
binary :: (Ord a) => [a] -> a -> Maybe Int
binary items item = binaryAux items item 0 (length items - 1)

ternaryAux :: (Ord a) => [a] -> a -> Int -> Int -> Maybe Int
ternaryAux [] _ _ _ = Nothing
ternaryAux items item l r
  | l > r = Nothing
  | item1 == item = Just mid1
  | item2 == item = Just mid2
  | item < item1 = binaryAux items item l (mid1 - 1)
  | item < item2 = binaryAux items item (mid1 + 1) (mid2 - 1)
  | item > item2 = binaryAux items item (mid2 + 1) r
  where
    delta = div (r - l) 3
    mid1 = l + delta
    mid2 = r - delta
    item1 = items !! mid1
    item2 = items !! mid2

ternary :: (Ord a) => [a] -> a -> Maybe Int
ternary items item = ternaryAux items item 0 (length items - 1)

File: Algos.jl/src/Searching.jl
module Searching

function linear(items::Vector, item)::Union{Nothing,Integer}
    for i in 1:length(items)
        if items[i] == item
            return i
        end
    end
    return nothing
end

export linear

function binary(items::Vector, item)::Union{Nothing,Integer}
    l = 1
    r = length(items)

    while l <= r
        m:: Integer = floor((l + r) / 2)
        if items[m] == item
            return m
        elseif item < items[m]
            r = m - 1
        else
            l = m + 1
        end
    end

    return nothing
end

export binary

function ternary(items::Vector, item)::Union{Nothing,Integer}
    l = 1
    r = length(items)

    while l <= r
        delta:: Integer = floor((r - l) / 3)
        m1 = l + delta
        m2 = r - delta
        if items[m1] == item
            return m1
        elseif items[m2] == item
            return m2
        elseif item < items[m1]
            r = m1 - 1
        elseif item < items[m2]
            l = m1 + 1
            r = m2 - 1
        else
            l = m2 + 1
        end
    end

    return nothing
end

export ternary

function kary(k::Int64, items::Vector, item)::Union{Nothing,Integer}
    l = 1
    r = length(items)

    while l <= r
        delta::Integer = floor((r - l) / (k + 1))
        lcur = copy(l)
        for m in map(x -> lcur + x * delta, (1: k + 1))
            if items[m] == item
                return m
            elseif items[m] < item
                l = m + 1
            else
                r = m - 1
                break
            end
        end
    end

    return nothing
end

export kary

end
